### ISSUE TITLE
Refactor and migrate AWS QA Integration Tests to Core repository

### ISSUE THREAD

**Issue by EduLeon12** [wazuh/wazuh#17788]
## Description 

AWS integration tests found in the QA repository will be moved to wazuh/wazuh repository.

We need to ensure the migration is done correctly and tests are properly running.

First, the current wazuh-qa testing framework needs to be migrated and refactored to adapt to the new changes.

Then the AWS ITs can be migrated to wazuh core repository. 

Also, tests should be checked for a possible update on fixtures, use cases, features, etc.
  
## Related issues

- [x] https://github.com/wazuh/qa-integration-framework/issues/36.
- [x] https://github.com/wazuh/wazuh/issues/19053.
- [x] https://github.com/wazuh/wazuh/issues/19052.
- [x] https://github.com/wazuh/wazuh-qa/issues/4714.
- [x] https://github.com/wazuh/wazuh/issues/22514
- [x] https://github.com/wazuh/wazuh/issues/20975.
- [x] https://github.com/wazuh/qa-integration-framework/issues/133
- [x] https://github.com/wazuh/qa-integration-framework/issues/150

## Branches

- Epic branch: [17788-aws-integration-test-migration-and-refactor](https://github.com/wazuh/wazuh/tree/17788-aws-integration-test-migration-and-refactor)
- [qa-integration-framework](https://github.com/wazuh/qa-integration-framework/tree/17788-migrate-aws-integration-tests).
- [wazuh](https://github.com/wazuh/wazuh/tree/17788-aws-its-migration).

**Comment by EduLeon12** [comment]
## Issue Update

Issue on hold until wazuh-qa credentials creation and assignment is completed.

**Comment by EduLeon12** [comment]
## Issue Update

First Run to tests credentials

```
root@ip-172-31-29-166:/wazuh-qa/tests/integration# pytest test_aws/
==================================== test session starts ====================================
platform linux -- Python 3.9.17, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /wazuh-qa/tests/integration, configfile: pytest.ini
plugins: testinfra-5.0.0, html-3.1.1, metadata-2.0.4
collected 198 items                                                   

test_aws/test_basic.py ................                        [  8%]
test_aws/test_discard_regex.py FFFFFFFFFFFFFFFFF               [ 16%]
test_aws/test_log_groups.py EF                                 [ 17%]
test_aws/test_only_logs_after.py EEEEEEEEEEEEEEEFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFxFE [ 41%]
test_aws/test_parser.py F........................F             [ 54%]
```

After reading some of the error and failure logs, it's probably related to credentials issues. 

For example:
```
botocore.exceptions.EndpointConnectionError: Could not connect to the endpoint URL: "https://wazuh-config-integration-tests.s3.amazonaws.com/AWSLogs/819751203818/Config/us-east-1/2023/7/11/819751203818_Config_us-east-1_ConfigHistory_AWS_20230711T1713Z_2284332328336210204.json"

/usr/local/lib/python3.9/dist-packages/botocore/httpsession.py:494: EndpointConnectionError
```

Next steps:

- [ ] Further, debug tests to conclude the reason for the failures.

**Comment by EduLeon12** [comment]
## Issue Update

Tests will need to be debugged or AWS client library needs to be tested locally to try each permit.
Because some of the buckets returning an "Access Denied" are listed in my resource policy.

This needs to be tested to ensure all permits are being handled correctly.

**Comment by EduLeon12** [comment]
## Issue update

Still working on debugging the Access Denied error. 

I have tried multiple ways to debug the error. Current AWS ITs cannot run on docker containers and VM (Vagrant) due to an unknown error on the framework installation.

Ec2 due to security reasons, I haven't been able to fully connect it to a Debugger, Will keep trying other methods.

**Comment by EduLeon12** [comment]
## Issue Update

After further debugging the error, I was able to get a more specific cause of the error:

```
ClientError('An error occurred (AccessDenied) when calling the ListObjectsV2 operation: Access Denied')
```

After reading about it it was found it's related to the API sync AWS does when calling the S3 buckets, the _CopyBucket_  action has been requested to be included in my IAM Role to keep testing the tests before migration

**Comment by EduLeon12** [comment]
## Issue Update

After Resolving permit issues for the IAM Role, almost all tests passed now.

Test module have been migrated to qa-integration-framework and test cases to wazuh/tests/integration.

Next steps:
- [ ] Re run tests when CreateRoleGroup has been granted to confirm IAM Role is working properly 
- [ ] Fix new import Errors when running test on Core repository.

**Comment by EduLeon12** [comment]
## Issue Update 

After debugging all Access Denied Errors, my role has been properly configured and is able to consistently run the AWS QA Integrations tests.

```
root@ip-172-31-85-239:/wazuh-qa/tests/integration# pytest --disable-warnings test_aws/
======================================== test session starts =========================================
platform linux -- Python 3.10.6, pytest-7.1.2, pluggy-1.2.0
rootdir: /wazuh-qa/tests/integration, configfile: pytest.ini
plugins: testinfra-5.0.0, html-3.1.1, metadata-3.0.0
collected 195 items

test_aws/test_basic.py ................                                                        [  8%]
test_aws/test_discard_regex.py ..............                                                  [ 15%]
test_aws/test_log_groups.py ..                                                                 [ 16%]
test_aws/test_only_logs_after.py .............................................x.               [ 40%]
test_aws/test_parser.py ..........................                                             [ 53%]
test_aws/test_path.py ..........................................                               [ 75%]
test_aws/test_path_suffix.py .........                                                         [ 80%]
test_aws/test_regions.py ........................                                              [ 92%]
test_aws/test_remove_from_bucket.py ...sss.........                                            [100%]

======================= 191 passed, 3 skipped, 1 xfailed in 3838.36s (1:03:58) =======================
```

Will continue working on imports error when trying to run the same tests after migrating them.

**Comment by EduLeon12** [comment]
## Issue Update

Aws module has been properly migrated to the framework, import errors fixed, and requirements updated. Currently Waiting on the permit to create a branch and PR to the qa-integration-framework.

While working in Wazuh test integration it was found a small refactor can be done to improve readability and avoid repeating code.

This will be done after ensuring the migration of the tests suite from wazuh-qa to wazuh repositories has been properly done. 

Next steps:
- [ ] Run tests found in `wazuh/test/integration/test_aws` successfully using `qa-integration-framework`.
- [ ] Refactor tests.

**Comment by EduLeon12** [comment]
## Issue Update

After attempting to run the test on a fresh installation (EC2) a lot of import errors and dependencies errors keep appearing due to previous environment having the requirements file found in wazuh-qa, that file includes dependencies not found in the qa-integration-framework.

A PR was created to migrate de aws module to the framework but it was changed to a draft until a fresh installation is able to properly run the tests.

**Comment by EduLeon12** [comment]
## Issue Update

For better visibility of the issue I propose to divide into 2 different subtask,
- One will be in charge of the migration only and make necessary changes to test configuration and to the qa-integration-framework to ensure tests are passing when ran.

- One to refactor current tests to avoid repeated code and enhance readibility

**Comment by EduLeon12** [comment]
## Issue Update

In the meantime migration still being worked.

Currently fixing the following error:

```
        # Check AWS module started
        log_monitor.start(
            timeout=session_parameters.default_timeout,
            callback=event_monitor.callback_detect_aws_module_start
>       ).result()
E       AttributeError: 'NoneType' object has no attribute 'result'

test_basic.py:103: AttributeError

```

This occurs because the log_monitor (FileMonitor) currently doesn't have a result method, this was eliminated in the new `qa-integration-framework`. The BaseMonitor now follows an Abstract Base Class design for the implementation.
A way to adapt the tests to ensure the behavior is the expect is needed.

**Comment by EduLeon12** [comment]
## Issue update

AWS Test has been migrated successfully. However, while running the entire suite of tests one error was found in one specific test of the _test_only_logs_after_ .
The reason for the failure is due to an error in the method _get_service_db_row(table_name)_ of the qa-integration-framework where the result of a query is _None_ when it shouldn't be, all attempts to replicate the error have been unsuccessful since it only shows when running the entire suit and not the mentioned test alone. 
This might hint at a race condition error but further debugging needs to be made to ensure the entire suite is consistently running correctly.

**Comment by EduLeon12** [comment]
## Issue Update 

After running the entire suit tests it is confirmed that the migration has been performed successfully. 

```
root@wazuh-master:/wazuh/tests/integration# pytest -x test_aws/ --disable-warnings
==================================== test session starts ====================================
platform linux -- Python 3.10.12, pytest-7.1.2, pluggy-1.2.0
rootdir: /wazuh/tests/integration, configfile: pytest.ini
plugins: testinfra-5.0.0, metadata-3.0.0, html-3.1.1
collected 195 items

test_aws/test_basic.py ................                                               [  8%]
test_aws/test_discard_regex.py ..............                                         [ 15%]
test_aws/test_log_groups.py ..                                                        [ 16%]
test_aws/test_only_logs_after.py .............................................x.      [ 40%]
test_aws/test_parser.py ..........................                                    [ 53%]
test_aws/test_path.py ..........................................                      [ 75%]
test_aws/test_path_suffix.py .........                                                [ 80%]
test_aws/test_regions.py ........................                                     [ 92%]
test_aws/test_remove_from_bucket.py ...sss.........                                   [100%]

============ 191 passed, 3 skipped, 1 xfailed, 7 warnings in 3723.08s (1:02:03) =============
```

**Comment by EduLeon12** [comment]
## Issue Update

As seen in Wazuh repository [PR](https://github.com/wazuh/wazuh/pull/18508), checks will fail until qa-integration-framework [PR](https://github.com/wazuh/qa-integration-framework/pull/25) is merged. That why the reason behind the _pending review_ state.

**Comment by Selutario** [comment]
ETA postponed to `12/09/2023`, reasons:
- One week for the review of the [PR to the qa-framework](https://github.com/wazuh/qa-integration-framework/pull/25) and application of changes.
- One week for the review of the [PR to wazuh/wazuh](https://github.com/wazuh/wazuh/pull/18508) and application of changes.

**Comment by EduLeon12** [comment]
## Issue Update

Due to the latest changes requested in [qa-integration-framework](https://github.com/wazuh/qa-integration-framework/pull/25) pull request, the issue has been upgraded to an epic.

All changes requiring an specific issue/branch will be merged to the epics branch respectively this being the current ones for each repository:

Branches:
- [qa-integration-framework](https://github.com/wazuh/qa-integration-framework/tree/17788-migrate-aws-integration-tests)
- [wazuh](https://github.com/wazuh/wazuh/tree/17788-aws-its-migration).

**Comment by TomasTurina** [comment]
Please, change base branch to 4.9.0.
